API.md

```diff
@@ -0,0 +1,121 @@
+API Proposal
+============
+## Use Cases
+- **Application Administration** (e.g. GPII): updating multiple settings of multiple (other) applications
+- **Application Settings** (e.g. my app): read/update/monitor applications's own settings
+
+
+## To throw or not to throw
+
+## Module Styles
+[Interesting read-up concerning nodejs module design](http://psteeleidem.com/node-module-factory-pattern/)
+
+### 1) Module style - Namespaced Collection of functions
```

Comment:
Hi @vilnius-leopold - happy to discuss further.
Here are my observations - firstly, that "being OO" doesn't clearly define a particular implementation idiom - for example, JS is prototype-based, uses GC and does not support weak references or destructors, which are notable differences to most classical implementations of OO in other languages. Whilst OO can be a perfectly valid technique when applied to a piece of user software in a single language, as an integration technology or low-level implementation strategy it is much more questionable. It's not an accident that, for example, as you observe in your API proposal, all of the node.js core APIs, e.g. "fs" are not object-based and are stateless wherever possible. Implementing APIs in a stateless way is almost always more reliable and efficient where it is possible (think of REST), and also increases the likelihood that they can be consumed in a cross-language and cross-architecture way. After all, bear in mind that C++ to this day has still been excluded from use within the Linux kernel. The node.js and core Linux devs are highly experienced people and have picked an appropriate implementation scheme for their infrastructure. And we are not writing a piece of user software, but instead a low-level API or integration technology like they are.

So this brings us to the issue with respect to gsettings. In fact, your reference to g_settings_new() has played right into my argument :) Since I think I observe, through studying the Gnome docs in detail, that our current usage of it is actually leaking memory. Unless I misunderstand, a GSettings object is a GObject, which must be freed after its use by a call to g_object_unref - my reference is this page: https://developer.gnome.org/gobject/stable/gobject-memory.html - let me know if you think this is accurate. This kind of doubt is exactly the kind of API blunder that I think exposing an unnecessarily object or instance-based API at a low level represents - the user of the API is exposed to the possibility of leaks and strange semantic issues that require their extreme care and detailed understanding to prevent.

We can make our node GSettings API much clearer by taking care of all such issues for our clients, and producing a stateless API that guarantees that at the end of an API call, there are no dangling resources that our clients need to pick up (unless this is absolutely necessary - for example, if in future we implement the "monitor" functionality).